library/modules/EventManager.cpp

```diff
@@ -58,11 +64,30 @@ static multimap<int32_t, EventHandler> tickQueue;
 
 //TODO: consider unordered_map of pairs, or unordered_map of unordered_set, or whatever
 static multimap<Plugin*, EventHandler> handlers[EventType::EVENT_MAX];
-static int32_t eventLastTick[EventType::EVENT_MAX];
+static std::unordered_map<EventHandler,int32_t> eventLastTick;
 
 static const int32_t ticksPerYear = 403200;
 
-void DFHack::EventManager::registerListener(EventType::EventType e, EventHandler handler, Plugin* plugin) {
+// this function is only used within the file in registerListener and manageTickEvent
+void enqueueTickEvent(EventHandler &handler){
+    int32_t when = 0;
+    df::world* world = df::global::world;
+    if ( world ) {
+        when = world->frame_counter + handler.freq;
+    } else {
+        if ( Once::doOnce("EventManager registerListener unhonored absolute=false") )
+            Core::getInstance().getConsole().print("EventManager::registerTick: warning! absolute flag=false not honored.\n");
+    }
+    handler.when = when;
+    tickQueue.emplace(handler.when, handler);
+}
+
+void DFHack::EventManager::registerListener(EventType::EventType e, EventHandler handler, Plugin* plugin, bool backlog) {
+    if(e == EventType::TICK){
+        enqueueTickEvent(handler);
+    }
+    int32_t tick = backlog ? -2 : (df::global::world ? df::global::world->frame_counter : -1);
+    eventLastTick[handler] = tick; // received ranges.. // backlog: -1,n // !world: 0,n // world: tick,n
     handlers[e].insert(pair<Plugin*, EventHandler>(plugin, handler));
 }
```

Comment:
Again, this can be a separate PR. This allows registering tick event listeners (instead of constantly re-queuing your tick events). Many of my onUpdate plugin functions could be replaced with this.